cpp/oneapi/dal/algo/svm/backend/gpu/working_set.hpp

```diff
@@ -0,0 +1,244 @@
+        n_vectors_ = n_vectors;
+        n_ws_ = n_ws;
+        sorted_f_inices_ =
+            pr::ndarray<std::uint32_t, 1>::empty(queue_, { n_vectors_ }, sycl::usm::alloc::device);
+        auto [indicator, indicator_event] =
+            pr::ndarray<std::uint32_t, 1>::zeros(queue_, { n_vectors_ }, sycl::usm::alloc::device);
+
+        if (!n_ws_) {
+            const std::int64_t max_wg_size = dal::backend::device_max_wg_size(queue_);
+            n_ws_ = std::min(dal::backend::down_pow2<std::uint32_t>(n_vectors_),
+                             dal::backend::down_pow2<std::uint32_t>(max_wg_size));
+        }
+        n_selected_ = 0;
+
+        values_sort_ =
+            pr::ndarray<Float, 1>::empty(queue_, { n_vectors_ }, sycl::usm::alloc::device);
+        buff_indices_ =
+            pr::ndarray<std::uint32_t, 1>::empty(queue_, { n_vectors_ }, sycl::usm::alloc::device);
+        ws_indices_ =
+            pr::ndarray<std::uint32_t, 1>::empty(queue_, { n_ws_ }, sycl::usm::alloc::device);
+
+        indicator_event.wait_and_throw();
+        indicator_ = indicator;
+    }
+
+    std::int64_t get_size() const {
+        return n_ws_;
+    }
+
+    const pr::ndarray<std::uint32_t, 1>& get_ws_indices() const {
+        return ws_indices_;
+    }
+
+    sycl::event copy_last_to_first(const dal::backend::event_vector& deps = {}) {
+        ONEDAL_ASSERT(ws_indices_.has_mutable_data());
+        const std::int64_t q = n_ws_ / 2;
+        std::uint32_t* ws_indices_ptr = ws_indices_.get_mutable_data();
+        auto copy_event =
+            dal::backend::copy(queue_, ws_indices_ptr, ws_indices_ptr + q, n_ws_ - q, deps);
+        n_selected_ = q;
+        return copy_event;
+    }
+
+    sycl::event reset_indicator_with_zeros(const pr::ndarray<std::uint32_t, 1>& idx,
+                                           pr::ndarray<std::uint32_t, 1>& indicator,
+                                           const std::int64_t n,
+                                           const dal::backend::event_vector& deps = {}) {
+        ONEDAL_ASSERT(idx.get_dimension(0) == n_ws_);
+        ONEDAL_ASSERT(indicator.get_dimension(0) == n_vectors_);
+        ONEDAL_ASSERT(indicator.has_mutable_data());
+
+        const std::uint32_t* idx_ptr = idx.get_data();
+        std::uint32_t* indicator_ptr = indicator.get_mutable_data();
+
+        const auto wg_size = std::min(dal::backend::propose_wg_size(queue_), n);
+        const auto range = dal::backend::make_multiple_nd_range_1d(n, wg_size);
+
+        auto reset_indicator_with_zeros_event = queue_.submit([&](sycl::handler& cgh) {
+            cgh.depends_on(deps);
+
+            cgh.parallel_for(range, [=](sycl::nd_item<1> item) {
+                const std::uint32_t i = item.get_global_id(0);
+                indicator_ptr[idx_ptr[i]] = 0;
+            });
+        });
+
+        return reset_indicator_with_zeros_event;
+    }
+
+    sycl::event select_ws(const pr::ndarray<Float, 1>& y,
```

Comment:
Not clear what this function does. The class is already named `working_set`, but this function according to its name selects working set.